### PR TITLE
Lifting sheaves from the basis of a distributive lattice

### PR DESCRIPTION
**Pull request by mzeuner**
This PR proves a "comparison lemma" for sheaves on distributive lattices: Given a presheaf on a basis of a distributive lattice taking values in an arbitrary category (with limits) and satisfying a sheaf property for bases, this can be extended to a sheaf on the whole lattice.

There seems to be quite a bit of overlap in the files [`Cubical.Categories.Presheaf.KanExtension`](https://github.com/agda/cubical/blob/master/Cubical/Categories/Presheaf/KanExtension.agda) and [`Cubical.Categories.Limits.RightKan`](https://github.com/agda/cubical/blob/master/Cubical/Categories/Limits/RightKan.agda) the latter being a bit more general. Eventually these should be merged or the results upstreamed, but maybe that's something for a different PR?

### CODE REVIEW

File: Cubical/Categories/DistLatticeSheaf/Extension.agda [hunk 1]
```diff
@@ -0,0 +1,1011 @@
...
+    -- the other direction, surprisingly hard
+    lemma3 : ∀ (c : ob C) (cc : Cone (funcComp F (BDiag (λ i → α i , α∈L' i))) c)
+               (f : C [ c , DLRan F .F-ob (⋁ α) ])
+           → isConeMor (lemma1 c cc) F[⋁α]Cone f
+           → isConeMor cc restCone f
+    lemma3 c cc f isConeMorF = isConeMorSingLemma _ cc restCone singCase
+      where
+      singCase : ∀ i → f ⋆⟨ C ⟩ coneOut restCone (sing i) ≡ coneOut cc (sing i)
+      singCase i =
+        (subst (λ g → f ⋆⟨ C ⟩ (F[⋁α]Cone .coneOut ((α i , α∈L' i) , ind≤⋁ α i)) ≡ g)
```
Comment (mortberg) [inline Cubical/Categories/DistLatticeSheaf/Extension.agda:362]: Redundant parentheses

File: Cubical/Categories/DistLatticeSheaf/Extension.agda [hunk 2]
```diff
@@ -0,0 +1,1011 @@
...
+                               (coneOut cc (sing i)) ccᵢSubst
+        ccᵢSubstFiller = subst-filler (λ x → C [ c , F .F-ob x ]) Σpathhelper (coneOut cc (sing i))
+
+        βSubstFiller : PathP (λ j → C [ c , F .F-ob (Σpathhelper2 j) ])
+                    (uniqβConeMor c cc (α i) (α∈L' i) (ind≤⋁ α i) .fst .fst)
+                    (coneOut (lemma1 c cc) ((α i , α∈L' i) , ind≤⋁ α i))
+        βSubstFiller = subst-filler (λ x → C [ c , F .F-ob x ]) Σpathhelper2
+                                    (uniqβConeMor c cc (α i) (α∈L' i) (ind≤⋁ α i) .fst .fst)
+
+        Σpathhelperpath : Σpathhelper2 ≡ sym Σpathhelper
```
Comment (mortberg) [inline Cubical/Categories/DistLatticeSheaf/Extension.agda:392]: ?

File: Cubical/Categories/DistLatticeSheaf/Extension.agda [hunk 3]
```diff
@@ -0,0 +1,1011 @@
...
+            where
+            B : I → Type ℓ
+            B = λ 𝕚 → (DLSubCat ^op) [ (Σpathhelper (~ 𝕚)) , (α i ∧l α j , β∈L' (α i) (α∈L' i) j) ]
+
+            path : coneOut cc (sing i) ⋆⟨ C ⟩ F .F-hom (≤m→≤j _ _ (∧≤RCancel _ _))
+                 ≡ coneOut cc (sing j) ⋆⟨ C ⟩ F .F-hom (≤m→≤j _ _ (∧≤LCancel _ _))
+            path with (i ≟Fin j)
+            ... | (lt i<j) = coneOutCommutes cc (singPairL {i<j = i<j})
+                           ∙ sym (coneOutCommutes cc singPairR)
+            ... | (gt j<i) = transp B2 i0 almostPath
```
Comment (mortberg) [inline Cubical/Categories/DistLatticeSheaf/Extension.agda:430]: transport, maybe eta expand B2?

File: Cubical/Categories/DistLatticeSheaf/Extension.agda [hunk 4]
```diff
@@ -0,0 +1,1011 @@
...
+      pbPr₂PathP i = DLRan F .F-hom
+                       (isProp→PathP {B = λ i → (⋁β≡x i) ≤ (⋁β++γ≡x∨y i)}
+                                     (λ _ → is-prop-valued _ _)
+                                     (subst (⋁ β ≤_) (sym (⋁Split++ β γ)) (∨≤RCancel _ _))
+                                     (hom-∨₁ L C x y) i)
+
+      squarePathP : PathP (λ i → pbPr₁PathP i ⋆⟨ C ⟩ cospanPath i .s₁
+                               ≡ pbPr₂PathP i ⋆⟨ C ⟩ cospanPath i .s₂)
+                          (pbCommutes ⋁Pullback) (Fsq L C x y (DLRan F))
+      squarePathP = toPathP (isSetHom C _ _ _ _)
```
Comment (mortberg) [inline Cubical/Categories/DistLatticeSheaf/Extension.agda:1011]: Combine with the other lemma

**Review by mortberg — CHANGES_REQUESTED** [review]

File: Cubical/Categories/DistLatticeSheaf/Base.agda [hunk 5]
```diff
@@ -125,7 +99,7 @@ module _ (L : DistLattice ℓ) (C : Category ℓ' ℓ'') (T : Terminal C) where
     Fsq F = F-square F sq
 
   isDLSheafPullback : (F : DLPreSheaf) → Type (ℓ-max (ℓ-max ℓ ℓ') ℓ'')
-  isDLSheafPullback F = (F-ob F 0l ≡ 𝟙)
+  isDLSheafPullback F = isTerminal C (F-ob F 0l)
                       × ((x y : L .fst) → isPullback C _ _ _ (Fsq x y F))
```
[the remark this answers is not in the document]
Comment (plt-amy) [reply Cubical/Categories/DistLatticeSheaf/Base.agda:104]: What, being terminal or being a pullback? Either way:
https://github.com/agda/cubical/blob/master/Cubical/Categories/Limits/Terminal.agda#L43 and https://github.com/agda/cubical/blob/master/Cubical/Categories/Limits/Pullback.agda#L42

[the remark this answers is not in the document]
Comment (mortberg) [reply Cubical/Categories/DistLatticeSheaf/Base.agda:104]: With the new definition this should be trivial

[the remark this answers is not in the document]
Comment (mzeuner) [reply Cubical/Categories/DistLatticeSheaf/Base.agda:97]: Added a proof of propositionality (?) for all the different definitions of sheaves.

File: Cubical/Categories/DistLatticeSheaf/Extension.agda [hunk 6]
```diff
@@ -0,0 +1,1011 @@
+{-# OPTIONS --safe --experimental-lossy-unification #-}
+module Cubical.Categories.DistLatticeSheaf.Extension where
+
+open import Cubical.Foundations.Prelude
```
[the remark this answers is not in the document]
Comment (mzeuner) [reply Cubical/Categories/DistLatticeSheaf/Extension.agda:4]: No :) removed all imports that were not used.

**Comment by mortberg** [comment]
Let me know when this is ready for another review/merging

File: Cubical/Categories/DistLatticeSheaf/Extension.agda (hunk 4 above)
Comment (mzeuner) [reply Cubical/Categories/DistLatticeSheaf/Extension.agda:1005]: Added a few more lines, so we can be absolutely sure we get the desired result...

**Comment by mzeuner** [comment]
Ok, managed to incorporate all the changes except for a more streamlined proof of `lemma3`.
Not sure if it's that big a problem, the proof will remain ugly and pretty much unreadable...

Should I rename `lemma1-4` though?

**Comment by mortberg** [comment]
> Ok, managed to incorporate all the changes except for a more streamlined proof of `lemma3`. Not sure if it's that big a problem, the proof will remain ugly and pretty much unreadable...
> 
> Should I rename `lemma1-4` though?

Yeah, that might be good. Especially if they are not private...

**Comment by mzeuner** [comment]
> Yeah, that might be good. Especially if they are not private...

I made lemma 1-3 private, as they're only used to prove lemma 4, which I now gave the more informative name `coverLemma`...

**Comment by mortberg** [comment]
> > Yeah, that might be good. Especially if they are not private...
> 
> I made lemma 1-3 private, as they're only used to prove lemma 4, which I now gave the more informative name `coverLemma`...

Great! Is this ready to be merged as well now?

**Comment by mzeuner** [comment]
> Great! Is this ready to be merged as well now?

Let me try one more thing...

File: Cubical/Categories/DistLatticeSheaf/Extension.agda (hunk 2 above)
Comment (mzeuner) [reply Cubical/Categories/DistLatticeSheaf/Extension.agda]: The problem is that they're not equal by refl, but I have an idea to drastically reduce transports.

**Pull request merged**